server/util/cache_metrics/cache_metrics.go

```diff
@@ -0,0 +1,175 @@
+package cache_metrics
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+
+	gstatus "google.golang.org/grpc/status"
+)
+
+// CacheTier is an enumerated label value that describes the level at
+// which the cache sits in the composable cache stack.
+//
+// This label enables Prometheus queries that don't break when the cache implementation
+// is changed. For example, to get cache metrics for the in-memory tier only (without having
+// to explicitly refer to "redis", "memcached", etc.), one could write this query:
+//
+//     sum by (backend) (rate(buildbuddy_cache_write_count{tier="memory"}[5m])
+//
+// These types of queries are useful when monitoring the cache because caches
+// at different tiers will have observed latencies at different orders of magnitude,
+// so each tier likely needs its own chart in order to be observed at an appropriate
+// resolution.
+type CacheTier string
+
+const (
+	// MemoryCacheTier is used for in-memory cache implementations, like memcached and redis.
+	MemoryCacheTier CacheTier = "memory"
+	// CloudCacheTier is used for cloud storage cache implementations, like GCS or AWS S3.
+	CloudCacheTier CacheTier = "cloud"
+)
+
+func MakeCacheLabels(tier CacheTier, backend string) prometheus.Labels {
+	return prometheus.Labels{
+		metrics.CacheTierLabel:    string(tier),
+		metrics.CacheBackendLabel: backend,
+	}
+}
+
+func appendLabels(a prometheus.Labels, b prometheus.Labels) prometheus.Labels {
+	labels := prometheus.Labels{}
+	for k, v := range a {
+		labels[k] = v
+	}
+	for k, v := range b {
+		labels[k] = v
+	}
+	return labels
+}
+
+// InstrumentedReader wraps a Reader and CacheTimer so that the embedded timer is stopped
+// and metrics are recorded on the last read.
+type InstrumentedReader struct {
+	io.Reader
+	timer         *CacheTimer
+	readSizeBytes int64
```

Comment:
s/readSizeBytes/totalBytesRead/g


I don't see where this value is incremented?
It should be inside of Read:

 n, err := ir.Reader.Read(p)
 totalBytesRead += n
 if err != nil {
   ir.timer.EndRead(totalBytesRead)
 }

Reply:
I set the size ahead of time, when constructing the `InstrumentedReader` in `NewInstrumentedReader` since we do the same thing for `Write`s. Do you think we should instead increment inside `Read` since we're overriding `Read` anyway?

Reply:
OK, got it.